phoenix-core/src/main/java/org/apache/phoenix/coprocessor/UncoveredGlobalIndexRegionScanner.java

```diff
@@ -0,0 +1,387 @@
+     *
+     * This implementation conforms to server paging such that if the server side operation takes
+     * more than pageSizeInMs, a dummy result is returned to signal the client that more work
+     * to do on the server side. This is done to prevent RPC timeouts.
+     *
+     * @param result
+     * @return boolean to indicate if there are more rows to scan
+     * @throws IOException
+     */
+    @Override
+    public boolean next(List<Cell> result) throws IOException {
+        long startTime = EnvironmentEdgeManager.currentTimeMillis();
+        boolean hasMore;
+        region.startRegionOperation();
+        try {
+            synchronized (innerScanner) {
+                if (state == State.READY && !indexRowIterator.hasNext()) {
+                    state = State.INITIAL;
+                }
+                if (state == State.INITIAL) {
+                    indexRows = new ArrayList();
+                    dataRows = Maps.newConcurrentMap();
+                    indexToDataRowKeyMap = Maps.newTreeMap(Bytes.BYTES_COMPARATOR);
+                    state = State.SCANNING_INDEX;
+                }
+                if (state == State.SCANNING_INDEX) {
+                    do {
+                        List<Cell> row = new ArrayList<Cell>();
+                        hasMore = innerScanner.nextRaw(row);
+                        if (!row.isEmpty()) {
+                            if (isDummy(row)) {
+                                result.addAll(row);
+                                return true;
+                            }
+                            lastIndexRowKey = CellUtil.cloneRow(row.get(0));
+                            indexToDataRowKeyMap.put(lastIndexRowKey, indexMaintainer.buildDataRowKey(
+                                    new ImmutableBytesWritable(lastIndexRowKey), viewConstants));
+                            indexRows.add(row);
+                            indexRowCount++;
+                            if (hasMore && (EnvironmentEdgeManager.currentTimeMillis() - startTime)
+                                    >= pageSizeMs) {
+                                getDummyResult(lastIndexRowKey, result);
+                                return true;
+                            }
+                        }
+                    } while (hasMore && indexRowCount < pageSizeInRows);
+                    state = State.SCANNING_DATA;
+                }
+                if (state == State.SCANNING_DATA) {
+                    scanDataTableRows(startTime);
+                    indexRowIterator = indexRows.iterator();
+                }
+                if (state == State.READY) {
+                    if (indexRowIterator.hasNext()) {
+                        List<Cell> indexRow = indexRowIterator.next();
+                        result.addAll(indexRow);
+                        try {
+                            IndexUtil.addTupleAsOneCell(result, new ResultTuple(dataRows.get(new ImmutableBytesPtr(
```

Comment:
Not very critical but we are assuming here that we will always find the data row corresponding to the index row. We can make the code more robust by handling the scenario where there is an extra verified row in the index so dataRows.get() will return null.

Reply:
Please note that the index table scan is not raw here. So we get only one valid version for a given index row. The corresponding data table row should exist. The only time it may not exist if the index table row is replicated but the data table row is not. So, if the data table row does not exist we should skip the index row. I will update the code.